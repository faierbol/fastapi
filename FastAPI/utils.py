from datetime import datetime
from typing import Optional, TypedDict

import jwt
from fastapi import Depends
from FastAPI import config
from FastAPI.config import ALGORITHM, database
from fastapi.security import OAuth2PasswordBearer
from pydantic import EmailStr

optional_login = OAuth2PasswordBearer(tokenUrl="auth/login", auto_error=False)
oauth2_schema = OAuth2PasswordBearer(tokenUrl="auth/login")


class JWTPayload(TypedDict):
    user_id: int
    sub: EmailStr
    exp: datetime


def decode_jwt(token: str) -> JWTPayload:
    return jwt.decode(token, config.SECRET_KEY, algorithms=[ALGORITHM])  # type: ignore


async def get_user_id(token: str) -> Optional[int]:
    query = "SELECT user_id FROM authtoken_token WHERE key = :key"
    row = await database.fetch_one(query=query, values={"key": token})
    return row["user_id"] if row else None


async def get_current_user_or_none(
    token: str = Depends(optional_login),
) -> Optional[int]:
    if token:
        return decode_jwt(token)["user_id"]
    else:
        return None


async def get_current_user(token: str = Depends(oauth2_schema)) -> int:
    return decode_jwt(token)["user_id"]
