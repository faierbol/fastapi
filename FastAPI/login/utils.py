from datetime import datetime, timedelta

import jwt
from FastAPI.config import ACCESS_TOKEN_EXPIRE_MINUTES, ALGORITHM, SECRET_KEY
from passlib.context import CryptContext

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


def create_access_token(subject: str, user_id: int) -> str:
    expire = datetime.utcnow() + timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    to_encode = {"exp": expire, "sub": str(subject), "user_id": user_id}
    encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
    return encoded_jwt


def verify_password(plain_password: str, hashed_password: str) -> bool:
    return pwd_context.verify(plain_password, hashed_password)  # type: ignore
