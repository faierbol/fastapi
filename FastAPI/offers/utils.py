from typing import Optional

from FastAPI.offers import crud
from FastAPI.offers.schemas import Status, ValidOffer
from pydantic import ValidationError


async def review_status_validator(offer_id: str) -> Optional[str]:
    offer = await crud.get_offer(offer_id) or {}
    try:
        ValidOffer(**offer)
        return None
    except ValidationError as e:
        return e.json()


async def set_review_status(offer_id: str) -> None:
    await crud.offer_set_status(offer_id=offer_id, status=Status.ACTIVE.value)
