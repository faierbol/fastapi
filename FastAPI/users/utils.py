import logging
import os.path
from datetime import date, datetime, timedelta
from typing import Any, Dict

import emails
import jwt
from emails.template import JinjaTemplate
from FastAPI import config
from FastAPI.config import ALGORITHM, SECRET_KEY
from passlib.context import CryptContext

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


def get_password_hash(password: str) -> str:
    return pwd_context.hash(password)  # type: ignore


def get_activation_jwt(email: str) -> str:
    delta = timedelta(hours=config.EMAIL_RESET_TOKEN_EXPIRE_HOURS)
    now = datetime.utcnow()
    expires = now + delta
    exp = expires.timestamp()
    return jwt.encode(
        {
            "exp": exp,
            "nbf": now,
            "sub": email,
        },
        key=SECRET_KEY,
        algorithm=ALGORITHM,
    )


def send_email(
    email_to: str,
    subject_template: str,
    html_template: str,
    environment: Dict[str, Any],
) -> None:
    message = emails.Message(
        subject=JinjaTemplate(subject_template),
        html=JinjaTemplate(html_template),
        mail_from=(config.EMAILS_FROM_NAME, config.EMAILS_FROM_EMAIL),
    )
    smtp_options = {"host": config.SMTP_HOST, "port": config.SMTP_PORT}
    if config.SMTP_TLS:
        smtp_options["tls"] = True
    if config.SMTP_USER:
        smtp_options["user"] = config.SMTP_USER
    if config.SMTP_PASSWORD:
        smtp_options["password"] = config.SMTP_PASSWORD
    response = message.send(to=email_to, render=environment, smtp=smtp_options)
    logging.info(f"send email result: {response}")


def send_new_account_email(
    email_to: str,
    username: str,
    activation_token: str,
) -> None:
    subject = f"Phoqer - New account for user {username}"
    with open(os.path.join(os.getcwd(), "email-templates/build/new_account.html")) as f:
        template_str = f.read()
    link = f"http://phoqer.com/api/v2/users/activation/{activation_token}"
    send_email(
        email_to=email_to,
        subject_template=subject,
        html_template=template_str,
        environment={
            "project_name": "Phoqer",
            "username": username,
            "email": email_to,
            "link": link,
            "registration_date": date.today().strftime("%d-%m-%Y"),
        },
    )
