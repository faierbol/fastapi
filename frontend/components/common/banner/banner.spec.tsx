import { render } from '../../../tests/tests.utils';
import Banner from './index';

describe('Banner component', () => {
    it('matches snapshot', () => {
        const { asFragment } = render(<Banner />, {});
        expect(asFragment()).toMatchSnapshot();
    });
});
