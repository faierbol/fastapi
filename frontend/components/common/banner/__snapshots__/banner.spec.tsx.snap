// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="root-0-2-2"
    type="button"
  >
    <h2
      class="title-0-2-4"
    >
      # Share with others and earn
    </h2>
    <p
      class="text-0-2-5"
    >
      Place an advertisement on any category and get additional income right now.
    </p>
  </button>
</DocumentFragment>
`;
